File: content/snippets/python/interview_lcci_16_03_intersection.py
from typing import List

EPS = 1e-9


def less_point(a: List[float], b: List[float]) -> bool:
    if abs(a[0] - b[0]) > EPS:
        return a[0] < b[0]
    return a[1] < b[1] - EPS


def cross(a: List[float], b: List[float], c: List[float]) -> float:
    return (b[0] - a[0]) * (c[1] - a[1]) - (b[1] - a[1]) * (c[0] - a[0])


def on_segment(a: List[float], b: List[float], p: List[float]) -> bool:
    return (
        abs(cross(a, b, p)) < EPS
        and min(a[0], b[0]) - EPS <= p[0] <= max(a[0], b[0]) + EPS
        and min(a[1], b[1]) - EPS <= p[1] <= max(a[1], b[1]) + EPS
    )


def intersection(start1: List[float], end1: List[float], start2: List[float], end2: List[float]) -> List[float]:
    if less_point(end1, start1):
        start1, end1 = end1, start1
    if less_point(end2, start2):
        start2, end2 = end2, start2

    d1 = cross(start1, end1, start2)
    d2 = cross(start1, end1, end2)
    d3 = cross(start2, end2, start1)
    d4 = cross(start2, end2, end1)

    if ((d1 > EPS and d2 > EPS) or (d1 < -EPS and d2 < -EPS) or
            (d3 > EPS and d4 > EPS) or (d3 < -EPS and d4 < -EPS)):
        return []

    a1 = end1[1] - start1[1]
    b1 = start1[0] - end1[0]
    c1 = a1 * start1[0] + b1 * start1[1]

    a2 = end2[1] - start2[1]
    b2 = start2[0] - end2[0]
    c2 = a2 * start2[0] + b2 * start2[1]

    det = a1 * b2 - a2 * b1
    if abs(det) < EPS:
        candidates = []
        if on_segment(start1, end1, start2):
            candidates.append(start2)
        if on_segment(start1, end1, end2):
            candidates.append(end2)
        if on_segment(start2, end2, start1):
            candidates.append(start1)
        if on_segment(start2, end2, end1):
            candidates.append(end1)
        if not candidates:
            return []
        candidates.sort(key=lambda p: (p[0], p[1]))
        return candidates[0]

    x = (b2 * c1 - b1 * c2) / det
    y = (a1 * c2 - a2 * c1) / det
    p = [x, y]
    return p if on_segment(start1, end1, p) and on_segment(start2, end2, p) else []


def main() -> None:
    print(intersection([0, 0], [1, 0], [1, 1], [0, -1]))


if __name__ == "__main__":
    main()

File: content/snippets/scala3/interview_lcci_16_03_intersection.scala
object Lcci1603Intersection:
    private val EPS = 1e-9

    private def lessPoint(a: Array[Double], b: Array[Double]): Boolean = {
        if (math.abs(a(0) - b(0)) > EPS) a(0) < b(0) else a(1) < b(1) - EPS
    }

    private def cross(a: Array[Double], b: Array[Double], c: Array[Double]): Double = {
        (b(0) - a(0)) * (c(1) - a(1)) - (b(1) - a(1)) * (c(0) - a(0))
    }

    private def onSegment(a: Array[Double], b: Array[Double], p: Array[Double]): Boolean = {
        math.abs(cross(a, b, p)) < EPS &&
        math.min(a(0), b(0)) - EPS <= p(0) && p(0) <= math.max(a(0), b(0)) + EPS &&
        math.min(a(1), b(1)) - EPS <= p(1) && p(1) <= math.max(a(1), b(1)) + EPS
    }

    def intersection(start1: Array[Double], end1: Array[Double], start2: Array[Double], end2: Array[Double]): Array[Double] =
        var s1 = start1.clone()
        var e1 = end1.clone()
        var s2 = start2.clone()
        var e2 = end2.clone()

        if (lessPoint(e1, s1)) {
            val t = s1; s1 = e1; e1 = t
        }
        if (lessPoint(e2, s2)) {
            val t = s2; s2 = e2; e2 = t
        }

        val d1 = cross(s1, e1, s2)
        val d2 = cross(s1, e1, e2)
        val d3 = cross(s2, e2, s1)
        val d4 = cross(s2, e2, e1)

        if (((d1 > EPS && d2 > EPS) || (d1 < -EPS && d2 < -EPS)) ||
            ((d3 > EPS && d4 > EPS) || (d3 < -EPS && d4 < -EPS))) {
            return Array.emptyDoubleArray
        }

        val a1 = e1(1) - s1(1)
        val b1 = s1(0) - e1(0)
        val c1 = a1 * s1(0) + b1 * s1(1)

        val a2 = e2(1) - s2(1)
        val b2 = s2(0) - e2(0)
        val c2 = a2 * s2(0) + b2 * s2(1)

        val det = a1 * b2 - a2 * b1
        if (math.abs(det) < EPS) {
            val candidates = scala.collection.mutable.ArrayBuffer[Array[Double]]()
            if (onSegment(s1, e1, s2)) candidates += s2
            if (onSegment(s1, e1, e2)) candidates += e2
            if (onSegment(s2, e2, s1)) candidates += s1
            if (onSegment(s2, e2, e1)) candidates += e1
            if (candidates.isEmpty) return Array.emptyDoubleArray
            val sorted = candidates.sortBy(p => (p(0), p(1)))
            return sorted.head
        }

        val x = (b2 * c1 - b1 * c2) / det
        val y = (a1 * c2 - a2 * c1) / det
        val p = Array(x, y)
        if (onSegment(s1, e1, p) && onSegment(s2, e2, p)) p else Array.emptyDoubleArray

    @main def main(): Unit =
        val ans = intersection(Array(0, 0), Array(1, 0), Array(1, 1), Array(0, -1))
        println(ans.mkString("[", ", ", "]"))
